### PR TITLE
Enable voxel layer

### PR DESCRIPTION
**Pull request by yathartha3**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (add tickets here #955 ) |
| Primary OS tested on | (Ubuntu 18.04) |
| Robotic platform tested on | Turtlebot Simulation |

---

## Description of contribution in a few bullet points
* Voxel layer enabled, and used pointcloud2 data.
* added short documentation in the costmap_2d readme file

voxel_obstacle_avoidance

## Future work that may be required in bullet points
* Test with real robot

### CODE REVIEW

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml [hunk 1]
```diff
@@ -0,0 +1,192 @@
+amcl:
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params_voxel.yaml]: This doesn’t need to be a new param file, just add the voxel layer to the existing yaml

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      global_frame: odom
+      plugin_names: ["voxel_layer", "inflation_layer"]
+      plugin_types: ["nav2_costmap_2d::VoxelLayer", "nav2_costmap_2d::InflationLayer"]
+      clearable_layers: ["voxel_layer"]
+      rolling_window: true
+      width: 3
+      height: 3
+      resolution: 0.05
+      robot_radius: 0.22
+      inflation_layer.cost_scaling_factor: 3.0
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params_voxel.yaml]: Inflation_layer. 

You should place that in an inflation layer namespace

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+      obstacle_layer:
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params_voxel.yaml]: There is no obstacle layer defined in this configuration. Remove please

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        enabled: True
+        pointcloud:
+          topic: /intel_realsense_r200_depth/points
+          max_obstacle_height: 2.0
+          clearing: True
+          marking: True
+          data_type: "PointCloud2"
+      voxel_layer:
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params_voxel.yaml]: Theres alot more important parameters that need to be here for a full definition, at minimum the resolution, min height, mark threshold. 

Check nav1 param files and make sure you’ve got the greatest hits of parameters.

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      always_send_full_costmap: True
+      observation_sources: pointcloud
+  local_costmap_client:
+    ros__parameters:
+      use_sim_time: True
+  local_costmap_rclcpp_node:
+    ros__parameters:
+      use_sim_time: True
+
+global_costmap:
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params_voxel.yaml]: See local costmap comments mirrored for global

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 6]
```diff
@@ -88,6 +88,8 @@ void VoxelLayer::onInitialize()
       "voxel_grid", custom_qos);
   }
 
+  voxel_pub_->on_activate();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp]: I see an on activate, but i dont see an on deactivate

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 7]
```diff
@@ -108,10 +110,10 @@ void VoxelLayer::matchSize()
 
 void VoxelLayer::reset()
 {
-  deactivate();
-  resetMaps();
+  ObstacleLayer::deactivate();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp]: Are these overloaded to not call these base functions?

**Comment by SteveMacenski** [comment]
Good stuff!

**Comment by SteveMacenski** [comment]
That’s 1 point in the tiny ticket sprint! Whack 2 more and you’re tied!

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml (hunk 4 above)
Comment (yathartha3) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: Okay I will put them here as well

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml (hunk 3 above)
Comment (yathartha3) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: voxel_layer also initializes obstacle_layer, so I believe obstacle_layer params are required. When I tried without specifying obstacle layer params, obstacles would not show up in the global costmap.

Comment (SteveMacenski) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: Thats a bug that needs to be fixed. 

Voxel layer _derives_ from the obstacle layer, but the base class should *not* have its parameters namespaced in the bass class. 

In this topic, one could make an argument about software practice, but the stronger argument is that a user never defined an obstacle layer, why must they add parameters for one? That’s confusing

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml (hunk 4 above)
Comment (SteveMacenski) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: I also like your intel realsense plug ;-) might want to make that d435 as the modern option

Comment (SteveMacenski) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: I'd think just `/points` if you wanted to make it really generic, but since its just an example configuration, if the TB3 configs have that as the topic configured for default testing in gazebo, I wouldn't argue with an intel realsense plug.

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml (hunk 1 above)
Comment (SteveMacenski) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: Yeah, it should just be one additional entry into the costmap layers name & plugin type vectors and the configuration thereafter

File: nav2_bringup/bringup/params/nav2_params_voxel.yaml (hunk 4 above)
Comment (yathartha3) [reply nav2_bringup/bringup/params/nav2_params_voxel.yaml]: The current documented TB3 model with depth camera had that topic name, and it would be easier to test out of the box. I am fine changing it though.

File: nav2_bringup/bringup/params/nav2_params.yaml [hunk 8]
```diff
@@ -140,10 +158,25 @@ global_costmap:
...
+          topic: /intel_realsense_r200_depth/points
+          max_obstacle_height: 2.0
+          clearing: True
+          marking: True
+          data_type: "PointCloud2"
       static_layer:
         map_subscribe_transient_local: True
       always_send_full_costmap: True
-      observation_sources: scan
+      observation_sources: scan pointcloud
```
Comment (SteveMacenski) [inline nav2_bringup/bringup/params/nav2_params.yaml]: These observation sources should be in each individual layer’s namespaces. This is what allows you to have multiple sensors processed in the same layer

File: nav2_costmap_2d/plugins/voxel_layer.cpp [hunk 9]
```diff
@@ -108,11 +110,12 @@ void VoxelLayer::matchSize()
 
 void VoxelLayer::reset()
 {
+  voxel_pub_->on_deactivate();
   deactivate();
   resetMaps();
-  voxel_grid_.reset();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/voxel_layer.cpp]: Why was this removed?

File: nav2_bringup/bringup/params/nav2_params.yaml (hunk 8 above)
Comment (SteveMacenski) [reply nav2_bringup/bringup/params/nav2_params.yaml]: Is this being properly processed as is? Or is it the defaults masking it? That’s a problem

File: nav2_costmap_2d/plugins/voxel_layer.cpp (hunk 9 above)
Comment (yathartha3) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: @SteveMacenski 
While executing the `reset()`, the `resetMaps()` calls `voxel_grid_.reset()`, so I thought it was redundant.

```
void VoxelLayer::reset()
{
  deactivate();
  resetMaps();
  voxel_grid_.reset();
  activate();
  undeclareAllParameters();
}

void VoxelLayer::resetMaps()
{
  Costmap2D::resetMaps();
  voxel_grid_.reset();
}
```

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/voxel_layer.cpp]: Looking at the git history in nav1 that appears to be the case

File: nav2_bringup/bringup/params/nav2_params.yaml (hunk 8 above)
Comment (yathartha3) [reply nav2_bringup/bringup/params/nav2_params.yaml]: Observation sources are now in individual layer's namespaces.

**Comment by SteveMacenski** [comment]
You’re not building, but fix all that and I’m happy.

**Comment by yathartha3** [comment]
> You’re not building, but fix all that and I’m happy.

I belive the build is failing due to something else. I was able to build and pass all checks on my machine using the lastest ros. Currently Navigation master is failing on Circle CI and Travis.

**Comment by SteveMacenski** [comment]
Oh, lovely

**Comment by SteveMacenski** [comment]
Did you show that you can add multiple sensor sources in the voxel (or obstacle) layer?

If so you can squash a few other tickets.

**Comment by yathartha3** [comment]
> Did you show that you can add multiple sensor sources in the voxel (or obstacle) layer?
> 
> If so you can squash a few other tickets in this same theme.

@SteveMacenski I tested with multiple sensors in the obstacle layer using param configuration gist shown below :
```
      plugin_names: ["obstacle_layer", "inflation_layer"]
      plugin_types: ["nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::InflationLayer"]
      obstacle_layer:
        enabled: True
        observation_sources: scan pointcloud
        scan:
          topic: /scan
          data_type: "LaserScan"
        pointcloud:
          topic: /intel_realsense_r200_depth/points
          data_type: "PointCloud2"
```
